modules/apps/data-engine/data-engine-rest-impl/src/main/java/com/liferay/data/engine/rest/internal/content/type/DataDefinitionContentTypeRegistry.java

```diff
@@ -54,61 +55,73 @@ public DataDefinitionContentType getDataDefinitionContentType(
 		String contentType) {
 
 		return Optional.ofNullable(
-			_dataDefinitionContentTypesByContentType.get(contentType)
+			_serviceTrackerMap.getService(contentType)
 		).orElseThrow(
 			() -> new DataDefinitionValidationException.MustSetValidContentType(
 				contentType)
 		);
 	}
 
-	@Reference(
-		cardinality = ReferenceCardinality.MULTIPLE,
-		policy = ReferencePolicy.DYNAMIC,
-		policyOption = ReferencePolicyOption.GREEDY
-	)
-	protected void addDataDefinitionContentType(
-		DataDefinitionContentType dataDefinitionContentType,
-		Map<String, Object> properties) {
-
-		if (!properties.containsKey("content.type")) {
-			return;
-		}
-
-		String contentType = MapUtil.getString(properties, "content.type");
-
-		_classNameIds.put(
-			contentType, dataDefinitionContentType.getClassNameId());
-
-		_dataDefinitionContentTypesByClassNameId.put(
-			dataDefinitionContentType.getClassNameId(),
-			dataDefinitionContentType);
-
-		_dataDefinitionContentTypesByContentType.put(
-			contentType, dataDefinitionContentType);
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_serviceTrackerMap = ServiceTrackerMapFactory.openSingleValueMap(
+			bundleContext, DataDefinitionContentType.class, null,
+			new ServiceTrackerCustomizer
+				<DataDefinitionContentType, DataDefinitionContentType>() {
+
+				@Override
+				public DataDefinitionContentType addingService(
```

Comment:
@JanisZhang I think it's better to use `ServiceTrackerMapListener` here, instead of using a `ServiceTrackerCustomizer`.